javascript_nodejs/50.contoso-cafe-bot/index.js

```diff
@@ -0,0 +1,88 @@
+// Note: Ensure you have a .env file and include botFilePath and botFileSecret.
+const ENV_FILE = path.join(__dirname, '.env');
+const env = require('dotenv').config({path: ENV_FILE});
+const BOT_CONFIGURATION_ERROR = 1;
+// Bot configuration section in the .bot file.
+// See https://aka.ms/about-bot-file to learn more about .bot file its use and bot configuration.
+const BOT_CONFIGURATION = 'contoso-cafe-bot';
+
+// Create HTTP server
+let server = restify.createServer();
+server.listen(process.env.port || process.env.PORT || 3980, function () {
+    console.log(`\n${server.name} listening to ${server.url}`);
+    console.log(`\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`);
+    console.log(`\nTo talk to your bot, open echoBot-with-counter.bot file in the Emulator`);
+});
+
+// .bot file path
+const BOT_FILE = path.join(__dirname, (process.env.botFilePath || ''));
+
+// Read bot configuration from .bot file. 
+let botConfig;
+try {
+    botConfig = BotConfiguration.loadSync(BOT_FILE, process.env.botFileSecret);
+} catch (err) {
+    console.log(`Error reading bot file. Please ensure you have valid botFilePath and botFileSecret set for your environment`);
+    process.exit(BOT_CONFIGURATION_ERROR);
+}
+
+// Get bot endpoint configuration by service name
+const endpointConfig = botConfig.findServiceByNameOrId(BOT_CONFIGURATION);
+
+// Create adapter. See https://aka.ms/about-bot-adapter to learn more about .bot file its use and bot configuration.
+const adapter = new BotFrameworkAdapter({
+    appId: endpointConfig.appId || process.env.microsoftAppID,
+    appPassword: endpointConfig.appPassword || process.env.microsoftAppPassword
+});
+
+// Define state store for your bot. See https://aka.ms/about-bot-state to learn more about bot state.
+const memoryStorage = new MemoryStorage();
+// CAUTION: The Memory Storage used here is for local bot debugging only. When the bot
+// is restarted, anything stored in memory will be gone. 
+// For production bots use Azure CosmosDB storage or Azure Blob storage providers. 
+// const { CosmosDbStorage } = require('botbuilder-azure');
+// const STORAGE_CONFIGURATION = 'CosmosDB'; 
+// const cosmosConfig = botConfig.findServiceByNameOrId(STORAGE_CONFIGURATION);
+// const cosmosStorage = new CosmosDbStorage({serviceEndpoint: cosmosConfig.connectionString, 
+//                                            authKey: ?, 
+//                                            databaseId: cosmosConfig.database, 
+//                                            collectionId: cosmosConfig.collection});
+
+// Create conversation state with in-memory storage provider. 
+const conversationState = new ConversationState(memoryStorage);
+
+// Create user state with in-memory storage provider. 
+const userState = new UserState(memoryStorage);
+
+// Register conversation state and user state as a middleware. 
+adapter.use(new BotStateSet(conversationState, userState));
```

Comment:
>BotStateSet [](start = 16, length = 11)

this is no longer a middleware.  

I think this should be
adapter.use(new AutoSaveStateMiddleware(conversationState, userState));